lib/swm050/gpio.c

```diff
@@ -0,0 +1,173 @@
+atomic pin setting.
+
+@param[in] gpios Unsigned int16. Pin identifiers @ref gpio_pin_id
+	     If multiple pins are to be changed, use bitwise OR '|' to separate
+	     them.
+*/
+void gpio_set(uint16_t gpios)
+{
+	GPIO_DATA |= gpios;
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief Clear a Group of Pins
+
+Set one or more pins of GPIO to 0. Please note that this chip doesn't support
+atomic pin setting.
+
+@param[in] gpios Unsigned int16. Pin identifiers @ref gpio_pin_id
+	     If multiple pins are to be changed, use bitwise OR '|' to separate
+	     them.
+*/
+void gpio_clear(uint16_t gpios)
+{
+	GPIO_DATA &= ~gpios;
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief Read a Group of Pins.
+
+@param[in] gpios Unsigned int16. Pin identifiers @ref gpio_pin_id
+	    If multiple pins are to be read, use bitwise OR '|' to separate
+	    them.
+@return Unsigned int16 value of the pin values. The bit position of the pin
+	value returned corresponds to the pin number.
+*/
+uint16_t gpio_get(uint16_t gpios)
+{
+	return GPIO_EXT & gpios;
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief Toggle a Group of Pins
+
+Toggle one or more pins of GPIO. The non-toggled pins are not affected.
+
+@param[in] gpios Unsigned int16. Pin identifiers @ref gpio_pin_id
+	     If multiple pins are to be changed, use bitwise OR '|' to separate
+	     them.
+*/
+void gpio_toggle(uint16_t gpios)
+{
+	uint32_t curr_status = GPIO_DATA & gpios;
+	GPIO_DATA = (GPIO_DATA & (~gpios)) | (~curr_status);
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief Set the direction of a Group of Pins
+
+Set the direction of one or more pins of GPIO.
+
+@param[in] gpios Unsigned int16. Pin identifiers @ref gpio_pin_id
+	     If multiple pins are to be changed, use bitwise OR '|' to separate
+	     them.
+@param[in] dir Unsigned int8. Pin direction @ref gpio_dir
+*/
+void gpio_dir(uint16_t gpios, uint8_t dir)
+{
+	if (dir == GPIO_INPUT)
+		GPIO_DIR &= ~gpios;
+	else if (dir == GPIO_OUTPUT)
+		GPIO_DIR |= gpios;
+}
+
+/*---------------------------------------------------------------------------*/
+/** @brief Select the alternative function of a Group of Pins
+
+Select the alternative function of one or more pins of GPIO.
+
+@param[in] gpios Unsigned int16. Pin identifiers @ref gpio_pin_id
+	     If multiple pins are to be changed, use bitwise OR '|' to separate
+	     them.
+@param[in] af_en Bool. Whether alternative function is selected
+*/
+void gpio_sel_af(uint16_t gpios, bool af_en)
+{
+	if (gpios & GPIO0)
+		GPIO_SEL = (GPIO_SEL & (~0x3)) | (af_en ? 0x1 : 0x0);
+	if (gpios & GPIO1)
+		GPIO_SEL = (GPIO_SEL & (~0xc)) | (af_en ? 0x4 : 0x0);
+	if (gpios & GPIO2)
+		GPIO_SEL = (GPIO_SEL & (~0x30)) | (af_en ? 0x10 : 0x0);
+	if (gpios & GPIO7)
+		GPIO_SEL = (GPIO_SEL & (~0xc000)) | (af_en ? 0x4000 : 0x0);
```

Comment:
do only thouse four pins support AF? (I've not looked at the manuals for this part, but that code looks odd)

Reply:
Yes, other pins can only do GPIO.

Although the datasheet is Chinese only, the pin setup picture has no Chinese inside, so I copied it here.

![image](https://user-images.githubusercontent.com/1916518/53230167-0ad79780-36c1-11e9-82b2-2fbeb81194a1.png)

Reply:
ok, cool, just the sort of thing thatlooks odd :)